crates/sui-adapter/src/adapter.rs

```diff
@@ -489,37 +487,33 @@ pub fn generate_package_id(
     modules: &mut [CompiledModule],
     ctx: &mut TxContext,
 ) -> Result<ObjectID, ExecutionError> {
-    let mut sub_map = BTreeMap::new();
     let package_id = ctx.fresh_id();
-    for module in modules.iter() {
-        let old_module_id = module.self_id();
-        let old_address = *old_module_id.address();
-        if old_address != AccountAddress::ZERO {
-            let handle = module.module_handle_at(module.self_module_handle_idx);
-            let name = module.identifier_at(handle.name);
+    let new_address = AccountAddress::from(package_id);
+
+    for module in modules.iter_mut() {
+        let self_handle = module.self_handle().clone();
+        let self_address_idx = self_handle.address;
+
+        let addrs = &mut module.address_identifiers;
+        let Some(address_mut) = addrs.get_mut(self_address_idx.0 as usize) else {
+            let name = module.identifier_at(self_handle.name);
             return Err(ExecutionError::new_with_source(
                 ExecutionErrorKind::PublishErrorNonZeroAddress,
-                format!("Publishing module {name} with non-zero address is not allowed"),
+                format!("Publishing module {name} with invalid address index"),
             ));
-        }
-        let new_module_id = ModuleId::new(
-            AccountAddress::from(package_id),
-            old_module_id.name().to_owned(),
-        );
-        if sub_map.insert(old_module_id, new_module_id).is_some() {
+        };
+
+        if *address_mut != AccountAddress::ZERO {
+            let name = module.identifier_at(self_handle.name);
             return Err(ExecutionError::new_with_source(
-                ExecutionErrorKind::PublishErrorDuplicateModule,
-                "Publishing two modules with the same ID",
+                ExecutionErrorKind::PublishErrorNonZeroAddress,
+                format!("Publishing module {name} with non-zero address is not allowed"),
             ));
-        }
-    }
+        };
 
-    // Safe to unwrap because we checked for duplicate domain entries above, and range entries are fresh ID's
-    let rewriter = ModuleHandleRewriter::new(sub_map).unwrap();
-    for module in modules.iter_mut() {
-        // rewrite module handles to reflect freshly generated ID's
-        rewriter.sub_module_ids(module);
+        *address_mut = new_address;
```

Comment:
IIRC, the reason I introduced bytecode_rewriter instead of doing this (which is indeed much simpler!) is to avoid changing the semantics of code that uses the constant address `0x0`. E.g., if you write the following code:

```
fun f(a: address) {
  if (a == 0x0) {
    // do something
  }
}
```

, I think this approach will effectively rewrite the code above to `if (a == <package_address>)`.

Reply:
The other edge case to worry about is when `address_identifiers` already contains the module address (which will cause the verifier to fail since it insists that `address_identifiers` creates no duplicates). E.g., if the derived package address is *A*, `address_identifiers` already contains *A*, and then this approach overwrites `0x0` with *A* and creates a duplicate.

I think this is basically impossible (I believe you can't construct such a program without finding a hash collision?) and is much less harmful than the problem above because the verifier will reject it, but worth mentioning.

Reply:
So I believe the case of address literals (i.e. constant expressions) is served out of the constant pool rather than the address identifier pool which is either amazing forethought or a great spot of luck! (@tnowacki does that sound right to you?)

I did a couple of sanity checks to that effect: First, I checked the opcodes, and none take an AddressIdentifierIndex, and second I compiled an example with an address literal, (it was one prefixed with an @ in case that's meaningfully different to a bare hexadecimal literal) and that address showed up in the ConstantPool and not the AddressIdentifierPool.

Reply:
Beautiful, thanks for checking this!

Reply:
> the address identifier pool which is either amazing forethought or a great spot of luck!

I'll go with forethought ;) 

In actuality I think I did it to simplify the code when deserializing/serializing them for ModuleIDs. It removed all sorts of weird edge cases around the formatting of the bytes. Additionally it was a space saving trick for not having to have a signature token ahead of each one. 
I think there was also just a conceptual divide that bothered me. We don't let random strings from the identifier pool become values, so why do that for addresses. Maybe we would want to change address identifiers someday without changing the value.

But let's go with forethought about these sorts of problems